### PR TITLE
Transformations

### PR DESCRIPTION
**Pull request by mfmezger**
Includes transformations for batch and single images.
transformtypes:
- Cropping
- Noise
- Blur

### CODE REVIEW

File: torchvtk/augmentation/DictTransform.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        self.noise_variance = noise_variance
+        DictTransform.__init__(self, self.device, apply_on=apply_on, dtype=dtype, batch_transform=batch_transform)
+
+    def transform(self, data):
+        #todo allow for batch
+        variance = random.uniform(self.noise_variance[0], self.noise_variance[1])
+        noise = np.random.normal(0.0, variance, size=data.shape)
+        noise = torch.from_numpy(noise)
+        noise = noise.to(self.device)
+        sample = torch.add(data, noise)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:74]:
```suggestion
        std = torch.rand(data.size(0) if data.ndim > 4 else 1, device=data.device, dtype=data.dtype)
        return data +  torch.randn_like(data) * std
```

File: torchvtk/augmentation/DictTransform.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        return data
+
+
+class NoiseDictTransform(DictTransform):
+
+    def __init__(self, device, noise_variance=(0.001, 0.05), apply_on=["vol"], dtype=torch.float32, batch_transform=False):
+
+        self.device = device
+        self.noise_variance = noise_variance
+        DictTransform.__init__(self, self.device, apply_on=apply_on, dtype=dtype, batch_transform=batch_transform)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:66]:
```suggestion
    def __init__(self, noise_variance=(0.001, 0.05), **kwargs):
        super().__init__(**kwargs)
        self.noise_variance = noise_variance  
```

File: torchvtk/augmentation/DictTransform.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        noise = np.random.normal(0.0, variance, size=data.shape)
+        noise = torch.from_numpy(noise)
+        noise = noise.to(self.device)
+        sample = torch.add(data, noise)
+        return sample
+
+
+class BlurDictTransform(DictTransform):
+
+    def __init__(self, apply_on=["vol"], dtype=torch.float32, device= "cpu", channels=1, kernel_size=(3, 3, 3), sigma=1, batch_transform=False):
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:80]: best use `**kwargs` for all params that go to DictTransform.
we'll have to find out how to still get proper documentation for the parameters with the kwargs

File: torchvtk/augmentation/DictTransform.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    def transform(self, data):
+        sample = data.unsqueeze(0)
+        vol = self.conv(sample, weight=self.weight, groups=self.groups, padding=1)
+        vol = vol.squeeze(0)
+        return vol
+
+
+class CroppingTransform(DictTransform):
+    def __init__(self, apply_on=["vol"], dtype=torch.float32, device= "cpu", size=20, position=0, batch_transform=False):
+        DictTransform.__init__(self, device=device, apply_on=apply_on, dtype=dtype, batch_transform=batch_transform)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:132]: **kwargs

File: torchvtk/augmentation/DictTransform.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        # todo check if that method of assigning works
+        # self.register_buffer('weight', kernel)
+        kernel = kernel.to(self.device)
+        self.weight = kernel
+        self.groups = self.channels
+
+        self.conv = f.conv3d
+
+    def transform(self, data):
+        sample = data.unsqueeze(0)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:124]: does this work as intended when using `batch_transform`? I would guess we get a 6D tensor then here. 
I suggest using `torchvtk.utils.make_5d` which does just that unless its already 5D

File: torchvtk/augmentation/DictTransform.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        # Reshape to depthwise convolutional weight
+        kernel = kernel.view(1, 1, *kernel.size())
+        kernel = kernel.repeat(self.channels, *[1] * (kernel.dim() - 1))
+
+        # todo check if that method of assigning works
+        # self.register_buffer('weight', kernel)
+        kernel = kernel.to(self.device)
+        self.weight = kernel
+        self.groups = self.channels
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:119]: what does the channels thing do? I suppose you want to blur each volume channel separately? I would suggest better getting the channels from the input tensor then. After `make_5d`, `tensor.size(1)` should always be correct

File: torchvtk/augmentation/DictTransform.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
+import math
+import random
+from abc import abstractmethod
+
+import numpy as np
+import torch
+from torch.nn import functional as f
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:7]: this is conventionally imported as uppercase `F`, I would stick to that

File: torchvtk/augmentation/DictTransform.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                tmp = self.transform(tmp)
+            else:
+                batch = []
+                for x in range(tmp.shape[0]):
+                    # get batch thing
+                    tmptensor = tmp[x, ...]
+                    # transform
+                    batch.append(self.transform(tmptensor))
+                # stack tensor back
+                tmp = torch.stack(batch, dim=0)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:48]: `batch_transform` should work the other way right? If its true, put in the whole thing.
I'm still not sure if this is the best place to take care of that. 

Maybe this `batch_transform` flag should rather be on the class level and should be hidden from the actual API? I mean each transform should be defined so it works with both 4D tensors and 5D tensors. This flag should then only split the batch if for example a rotation with different angles can't be easily parallelized over the whole batch. The user however should never have to decice wether a transform is applied on a batch or single item. The implementation of the transforms should take care of this. 

I'm suggesting we keep this batch_transform flag, but set it fixed in each transform's `__init__`.

**Review by xeTaiz — CHANGES_REQUESTED** [review]
I have some comments, but looks good overall. I would suggest using PyTorch where ever possible

File: torchvtk/augmentation/DictTransform.py (hunk 2 above)
Comment (mfmezger) [reply torchvtk/augmentation/DictTransform.py]: resolved.

File: torchvtk/augmentation/DictTransform.py (hunk 8 above)
Comment (mfmezger) [reply torchvtk/augmentation/DictTransform.py]: I agree that it is not very elegant. i removed it completely and changed the code so that it does not matter. it is tested with 4D and 5D tensor. Not tested with 2D or 3D tho. But that should not be a problem, because its kinda out of scope or?

File: torchvtk/augmentation/DictTransform.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            if self.dtype is torch.float16:
+                tmp = tmp.to(torch.float16)
+            elif self.dtype is not torch.float16 or torch.float32:
+                tmp = tmp.to(torch.float32)
+            tmp = tmp.to(self.device)
+            data[key] = tmp
+        return data
+
+
+class NoiseDictTransform(DictTransform):
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:59]: I just wonder should this have a clip parameter? Usually you would assume the data to be in a range, e.g. `[0,1]` and using a normal distribution can generate outliers from this. 
Should we enable this option or just implement a Clip Transform or so?

File: torchvtk/augmentation/DictTransform.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    Transformations for adding noise to images.
+    """
+
+    def __init__(self, noise_variance=(0.001, 0.05), **kwargs):
+        """
+
+        :param noise_variance: The variance of the noise added to the  image.
+        :param kwargs: Arguments of the super class.
+        """
+        self.device = kwargs["device"]
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:70]: self.device is not needed

File: torchvtk/augmentation/DictTransform.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        self.noise_variance = noise_variance
+        DictTransform.__init__(self, **kwargs)
+
+    def transform(self, data):
+        std = torch.rand(data.size(0) if data.ndim > 5 else 1, device=data.device, dtype=data.dtype)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:75]: ndim > 4. it will never be > 5

File: torchvtk/augmentation/DictTransform.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        kernel = kernel.view(1, 1, *kernel.size())
+        kernel = kernel.repeat(self.channels, *[1] * (kernel.dim() - 1))
+
+        kernel = kernel.to(self.device)
+        self.weight = kernel
+        self.conv = F.conv3d
+
+    def transform(self, data):
+        vol = make_5d(data)
+        vol = self.conv(vol, weight=self.weight, groups=self.channels, padding=1)
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:121]: For blurring we would need to determine the amount using `padding = kernel_size // 2`.
Also usually replication padding is used when doing a gaussian blur, since zero padding creates a gradient at the borders. Assuming `self.pad = kernel_size // 2` I would suggest
```suggestion
        vol = F.pad(make_5d(data), (self.pad, )*6, mode='replicate')
        return self.conv(vol, weight=self.weight, groups=self.channels, padding=0)
```

File: torchvtk/augmentation/DictTransform.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        :param position: Middle point of the cropped region.
+        :param kwargs: Arguments for super class.
+        """
+        DictTransform.__init__(self, **kwargs)
+        self.size = size
+        self.position = position
+
+    def transform(self, data):
+        data = self.get_center_crop(data, self.size)
+        return data
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:142]:
```suggestion
        return self.get_center_crop(data, self.size)
```

Removing temporary variables can actually give significant speedup I think.

**Review by xeTaiz — CHANGES_REQUESTED** [review]
Just some more minor things, looks ready to merge :) 

About the clip comment: I will add this to the transformation list issue, i don't think its necessary for this PR

File: torchvtk/augmentation/DictTransform.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+                tmp = tmp.to(torch.float32)
+
+            tmp = self.transform(tmp)
+
+            if self.dtype is torch.float16:
+                tmp = tmp.to(torch.float16)
+            elif self.dtype is not torch.float16 or torch.float32:
+                tmp = tmp.to(torch.float32)
+            tmp = tmp.to(self.device)
+            data[key] = tmp
```
Comment (xeTaiz) [inline torchvtk/augmentation/DictTransform.py:56]: This can be compacted.
Assuming tmp is a tensor, no checks for device and dtype are necessary since the call to `to` is a noop if the type or device are already correct. 
if tmp is a numpy array its converted to pytorch. Cant do this on `not torch.is_tensor` alone.
if tmp is something completely different, which we might want to support (maybe adding a transform to a transfer function or some keypoints or whatever), just rely on the transform to handle it appropriately (For now those would only be custom implemented transforms, so the user would probably know)

```suggestion
            tmp = data[key]
            isinstance(tmp, np.ndarray): # Convert from NumPy
                tmp = torch.from_numpy(tmp)   
                
            if torch.is_tensor(tmp): # If tmp is tensor, control type and device
                data[key] = self.transform(tmp.to(self.dtype).to(self.device))
            else: 
                data[key] = self.transform(tmp)
```

**Review by xeTaiz — APPROVED** [review]
